### PR TITLE
2693 feature/bsync support v2

### PR DESCRIPTION
**Pull request by Ryoken**
#### What's this PR do?
Adds support for all V2.x buildingsync files

#### How should this be manually tested?
Upload a version 2.3.0 buildingsync file (such as <a download href=https://buildingenergyscore.energy.gov/documents/example_SF_audit_report_BS_v2.3_033121.xml>this one</a>)
#### What are the relevant tickets?
https://github.com/SEED-platform/seed/issues/2693

### CODE REVIEW

File: seed/building_sync/building_sync.py [hunk 1]
```diff
@@ -365,15 +369,19 @@ def _parse_version(self):
...
             raise ParsingError('Expected BuildingSync element as root element in xml')
 
+        # first check for a version attribute in the buldingsync tag
+        if "version" in bsync_element.attrib:
+            return bsync_element.attrib["version"]
+
+        # second check if it's a file form Audit Template Tool and infer the version
+        # Currently ATT doesn't include a schemaLocation so this is necessary
+        if self._is_from_audit_template_tool():
+            return self.BUILDINGSYNC_V_LATEST
```
Comment (macintoshpie) [inline seed/building_sync/building_sync.py:383]: Ah I didn't think about this, but if a file is missing the version attribute and it appears to be from ATT, then we should assume the bsync version is `2.0`, not the latest (v2.0 was the last version which didn't require a version attribute). Could you update the comment to say something like
```
# second check if it's a file form Audit Template Tool
# if so, then it must be a 2.0 file b/c that was the last version which didn't require @version
```

File: seed/building_sync/validation_client.py [hunk 2]
```diff
@@ -13,16 +13,11 @@ class ValidationClientException(Exception):
...
-    files = [
-        ('file', file_)
-    ]
-
     return requests.request(
         "POST",
         VALIDATION_API_URL,
-        data=payload,
-        files=files,
+        data={'schema_version': schema_version},
```
Comment (macintoshpie) [inline seed/building_sync/validation_client.py:19]: Ugh, I just found out that if someone uploads a v2.0 file, we will pass a version of 2.0 (which is correct according to our tags) but selection tool requires 2.0.0. Can we add a check in this function to change version to 2.0.0 if someone passes BUILDINGSYNC_V2_0?

File: seed/data_importer/tasks.py [hunk 3]
```diff
@@ -1620,12 +1621,15 @@ def pair_new_states(merged_property_views, merged_taxlot_views):
 def _validate_use_cases(file_pk, progress_key):
     import_file = ImportFile.objects.get(pk=file_pk)
     progress_data = ProgressData.from_key(progress_key)
+    bs = BuildingSync()
+    bs.import_file(import_file.file)
```
Comment (macintoshpie) [inline seed/data_importer/tasks.py:1625]: Yet another gotcha, we allow users to upload zip files of BuildingSync files, so this causes the BuildingSync class to try and parse a zip. We might have to open the zip, get the version of the first file and validate them all as that version... What do you think about that? Alternatively we could iterate through all the files and validate them individually

File: seed/building_sync/validation_client.py [hunk 4]
```diff
@@ -13,16 +13,11 @@ class ValidationClientException(Exception):
...
+        files={'file': (file_.name, open(file_.name, 'r').read(), 'application/xml')},
```
Comment (macintoshpie) [inline seed/building_sync/validation_client.py:20]: We also allow folks to upload zips (and validate zips too). I didn't get the chance to test this yet, but I'm wondering if specifying this content-type will work.

Here's an example zip we use: seed/building_sync/tests/data/ex_1_and_buildingsync_ex01_measures.zip

File: seed/building_sync/building_sync.py [hunk 5]
```diff
@@ -43,9 +43,13 @@ class ParsingError(Exception):
 class BuildingSync(object):
     BUILDINGSYNC_V2_0 = '2.0'
     BUILDINGSYNC_V2_2_0 = '2.2.0'
+    BUILDINGSYNC_V2_3_0 = '2.3.0'
```
Comment (macintoshpie) [inline seed/building_sync/building_sync.py:46]: Not sure why I didn't add this earlier, but could you add 2.1.0 as well?

**Review by macintoshpie — CHANGES_REQUESTED** [review]
Great job so far, glad this is looking like it's not requiring many big changes! Just a few comments about tweaks we'll have to make (sorry for more gotchas).

Could you also add these new schemas to the repo? We store them here:
seed/building_sync/schemas/

We should probably have some test which asserts there's a schema for every version we claim to support, but don't feel like you have to add that now.

These schema files are used when exporting files to make sure we structure the file correctly.

File: seed/building_sync/building_sync.py (hunk 1 above)
Comment (Ryoken) [reply seed/building_sync/building_sync.py:383]: gotcha!

File: seed/building_sync/validation_client.py (hunk 2 above)
Comment (Ryoken) [reply seed/building_sync/validation_client.py:19]: I instead just updated BUILDINGSYNC_V2_0 to be "2.0.0" as I don't believe it's ever used to verify "2.0".  Can you see any problems with this?

File: seed/data_importer/tasks.py (hunk 3 above)
Comment (Ryoken) [reply seed/data_importer/tasks.py:1625]: I honestly feel we should validate all the files if a ZIP is presented.  I'll refactor to accomodate!

File: seed/building_sync/validation_client.py (hunk 4 above)
Comment (Ryoken) [reply seed/building_sync/validation_client.py:20]: Ah, I had specified content type in an attempt to debug the file upload issue.  I don't think it's needed in either case however so I'll just remove it.

File: seed/building_sync/validation_client.py (hunk 2 above)
Comment (macintoshpie) [reply seed/building_sync/validation_client.py:19]: I'm worried that might cause some issues, our official tag is [v2.0](https://github.com/BuildingSync/schema/releases/download/v2.0/BuildingSync.xsd) (not 2.0.0) which people would use in schemaLocation to reference the schema, which we were parsing out to determine the BSync version originally (before v2.1 added the `@version` attribute).

File: seed/data_importer/tasks.py [hunk 6]
```diff
@@ -1620,12 +1622,33 @@ def pair_new_states(merged_property_views, merged_taxlot_views):
...
+                with tempfile.TemporaryDirectory() as temp_dir:
+                    zip_file.extractall(path=temp_dir)
+                    for file_name in zip_file.namelist():
+                        bs = BuildingSync()
+                        bs.import_file(f'{temp_dir}/{file_name}')
+                        if found_version == 0:
+                            found_version = bs.version
+                        elif found_version != bs.version:
+                            raise Exception(f'Zip contains multiple BuildingSync versions (found {found_version} and {bs.version})')
+            import_file.refresh_from_db()
```
Comment (macintoshpie) [inline seed/data_importer/tasks.py:1641]: Nice! worked for me!

**Review by macintoshpie — APPROVED** [review]
Thanks! This looks great, I tweaked two things:
- added a migration to rename the default BuildingSync column mapping to just v2 instead of v2.0 now that we are importing more bsync versions
- fixed exports for other bsync file versions by tweaking get_schema method

If you could take a look at my changes and test it yourself by running the migration and importing / exporting a v2.3 bsync file, then I think this will be good to go!

Imported 2.3 file
Screen Shot 2021-08-05 at 8 02 49 AM
Screen Shot 2021-08-05 at 8 02 55 AM

Error when trying to import a zip with multiple versions of BSync (nice!)
Screen Shot 2021-08-05 at 8 06 36 AM

**Comment by Ryoken** [comment]
Works for me!  Thanks for the additions.

**Pull request merged**